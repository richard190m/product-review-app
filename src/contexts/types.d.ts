import { ReactNode } from 'react';

export type ReviewProviderProps = {
  children: ReactNode;
};
