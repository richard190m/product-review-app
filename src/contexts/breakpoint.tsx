import { BreakpointContextInterface, BreakpointProviderProps } from './types';
import { createContext, useEffect, useState } from 'react';

export const BreakpointContext =
  createContext<BreakpointContextInterface | null>(null);

export const BreakpointProvider = ({ children }: BreakpointProviderProps) => {
  const [breakpoint, setBreakpoint] = useState<string>('');

  const handleBreakpoint = () => {
    if (window.innerWidth < 768) breakpoint != 'xs' && setBreakpoint('xs');
    else if (window.innerWidth < 1024)
      breakpoint != 'sm' && setBreakpoint('sm');
    else if (window.innerWidth < 1440)
      breakpoint != 'md' && setBreakpoint('md');
    else breakpoint != 'lg' && setBreakpoint('lg');
  };

  useEffect(() => {
    handleBreakpoint();

    function handleResize() {
      handleBreakpoint();
    }

    window.addEventListener('resize', handleResize);

    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return (
    <BreakpointContext.Provider value={{ breakpoint }}>
      {children}
    </BreakpointContext.Provider>
  );
};
