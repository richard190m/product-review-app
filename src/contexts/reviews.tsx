import { createContext, useEffect, useState } from 'react';
import {
  HandleSubmit,
  Review,
  ReviewProviderProps,
  ReviewsContextInterface,
} from './types';
import MockReviews from '../mocks/MockReviews';

export const ReviewsContext = createContext<ReviewsContextInterface | null>(
  null,
);

export const ReviewsProvider = ({ children }: ReviewProviderProps) => {
  const [reviews, setReviews] = useState<Review[]>([]);

  const handleSubmit: HandleSubmit = review => {
    setReviews([...reviews, review]);
  };

  useEffect(() => {
    // Load Reviews (mocks)
    setReviews(MockReviews);
  }, []);

  return (
    <ReviewsContext.Provider value={{ reviews, submit: handleSubmit }}>
      {children}
    </ReviewsContext.Provider>
  );
};
