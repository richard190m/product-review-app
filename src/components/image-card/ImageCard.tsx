import { Container, Img } from './styles';
import React, { useContext, useEffect, useState } from 'react';
import { BreakpointContext } from '../../contexts/breakpoint';
import imageSizes from '../../utils/imageSizes';

const ImageCarousel = () => {
  const breakpointContext = useContext(BreakpointContext);
  const [productImageUrl, setProductImageUrl] = useState('');

  useEffect(() => {
    if (breakpointContext?.breakpoint) {
      setProductImageUrl(
        `https://source.unsplash.com/random/${
          imageSizes[breakpointContext.breakpoint]
        }/?shoes`,
      );
    }
  }, [breakpointContext?.breakpoint]);

  return (
    <Container>
      <Img url={productImageUrl} />
    </Container>
  );
};

export default ImageCarousel;
