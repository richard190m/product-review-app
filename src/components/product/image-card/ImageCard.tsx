import { Brand, Container, Img } from './styles';
import React, { useContext, useEffect, useState } from 'react';
import { BreakpointContext } from '../../../contexts/breakpoint';
import { ImageCardProps } from './types';
import imageSizes from '../../../utils/imageSizes';
import PropTypes from 'prop-types';

const ImageCarousel = ({ brand, alt }: ImageCardProps) => {
  const breakpointContext = useContext(BreakpointContext);
  const [productImageUrl, setProductImageUrl] = useState('');

  useEffect(() => {
    if (breakpointContext?.breakpoint) {
      setProductImageUrl(
        `https://source.unsplash.com/random/${
          imageSizes[breakpointContext.breakpoint]
        }/?shoes`,
      );
    }
  }, [breakpointContext?.breakpoint]);

  return (
    <Container>
      <Img src={productImageUrl} alt={alt} />
      <Brand>{brand}</Brand>
    </Container>
  );
};

ImageCarousel.propTypes = {
  brand: PropTypes.string,
  alt: PropTypes.string.isRequired,
};

export default ImageCarousel;
