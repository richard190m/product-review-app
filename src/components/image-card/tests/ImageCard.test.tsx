import { BreakpointProvider } from '../../../contexts/breakpoint';
import ImageCard from '../ImageCard';
import { render } from '@testing-library/react';

describe('ImageCard', () => {
  test('should render correctly', () => {
    const { container } = render(
      <BreakpointProvider>
        <ImageCard />
      </BreakpointProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
