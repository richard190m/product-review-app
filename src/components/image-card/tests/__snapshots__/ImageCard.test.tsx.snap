// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCard should render correctly 1`] = `
<div>
  <div
    class="sc-bczRLJ dQQVWG"
  >
    <div
      class="sc-gsnTZi kgKmUF"
    />
  </div>
</div>
`;
